unit: Can not multiply dimensionless units

<!--
Please make sure your issue title matches the Go convention; a summary
of the problem, prefixed by the primary affected package.
-->
### What are you trying to do?
Multiply unit.Dimless by a custom unit.

### What did you do?
<!--
Please include a link to a minimal reproducer here. Preferably this should
be a complete runnable example at https://play.golang.org/.
-->
https://play.golang.org/p/n17wR5TL_hx

### What did you expect to happen?
No panic with viable result, dimensions equal to other unit.

### What actually happened?
panic: assignment to entry in nil map

### What version of Go and Gonum are you using?
<!--
Paste the output of `go version` and if you are installing Gonum from source, paste
the output of `(cd $(go env GOPATH)/src/gonum.org/v1/gonum && git rev-parse HEAD)`.
If you are using modules, also paste the output of `grep gonum.org/v1/gonum go.sum`,
executed in the root of your dependent module.
-->
`go version go1.12.1 linux/amd64`
```
gonum.org/v1/gonum v0.0.0-20190409070159-6e46824336d2 h1:IZ343hUHFQ/v9IrSpcT1ih//e8VIctin7w2WyVl4kcc=
gonum.org/v1/gonum v0.0.0-20190409070159-6e46824336d2/go.mod h1:2ltnJ7xHfj0zHS40VVPYEAAMTa3ZGguvHGBSJeRWqE0=
```


### Does this issue reproduce with the current master?

Yes

### Proposed fix
https://github.com/gonum/gonum/blob/master/unit/unittype.go#L171
Should be `return Dimensions{}` rather than `return nil`.

--- a/unit/unit_test.go
+++ b/unit/unit_test.go
@@ -159,6 +159,47 @@ func TestDimensionEquality(t *testing.T) {
 	}
 }
 
+var operationTests = []struct {
+	recvOp func(Uniter) *Unit
+	param  Uniter
+	want   Uniter
+}{
+	{Dimless(1).Unit().Add, Dimless(2), Dimless(3)},
+	{Dimless(1).Unit().Mul, Dimless(2), Dimless(2)},
+	{Dimless(1).Unit().Mul, Length(2), Length(2)},
+	{Length(1).Unit().Mul, Dimless(2), Length(2)},
+	{Dimless(1).Unit().Div, Length(2), New(0.5, Dimensions{LengthDim: -1})},
+	{Length(1).Unit().Div, Dimless(2), Length(0.5)},
+}
+
+func TestOperations(t *testing.T) {
+	for i, test := range operationTests {
+		var got Uniter
+		if panics(func() { got = test.recvOp(test.param) }) {
+			t.Errorf("unexpected panic for test %d", i)
+			continue
+		}
+		if !DimensionsMatch(got, test.want) {
+			t.Errorf("dimension mismatch for test %d: got=%v want=%v", i, got.Unit(), test.want.Unit())
+		}
+		if got.Unit().Value() != test.want.Unit().Value() {
+			t.Errorf("value mismatch for test %d: got=%v want=%v", i, got.Unit().Value(), test.want.Unit().Value())
+		}
+	}
+}
+
+// panics returns true if the function panics.
+func panics(f func()) (b bool) {
+	defer func() {
+		err := recover()
+		if err != nil {
+			b = true
+		}
+	}()
+	f()
+	return
+}
+
 type UnitStructer interface {
 	UnitStruct() *UnitStruct
 }

--- a/unit/unittype.go
+++ b/unit/unittype.go
@@ -167,7 +167,7 @@ var (
 type Dimensions map[Dimension]int
 
 func (d Dimensions) clone() Dimensions {
-	if len(d) == 0 {
+	if d == nil {
 		return nil
 	}
 	c := make(Dimensions, len(d))

